core/src/main/java/com/graphhopper/routing/util/MaxSpeedCalculator.java

```diff
@@ -0,0 +1,153 @@
+package com.graphhopper.routing.util;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.graphhopper.routing.ev.*;
+import com.graphhopper.routing.util.parsers.helpers.OSMValueExtractor;
+import com.graphhopper.storage.Graph;
+import de.westnordost.osm_legal_default_speeds.LegalDefaultSpeeds;
+import de.westnordost.osm_legal_default_speeds.RoadType;
+import de.westnordost.osm_legal_default_speeds.RoadTypeFilter;
+import kotlin.jvm.functions.Function0;
+import kotlin.jvm.functions.Function2;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class MaxSpeedCalculator {
+
+    public static void fillMaxSpeed(Graph graph,
+                                    EnumEncodedValue<UrbanDensity> urbanDensityEnc,
+                                    EnumEncodedValue<RoadClass> roadClassEnc,
+                                    BooleanEncodedValue roadClassLinkEnc,
+                                    EnumEncodedValue<Country> countryEnumEncodedValue,
+                                    DecimalEncodedValue maxSpeedEnc) {
+
+        SpeedLimitsJson data;
+        try {
+            data = new ObjectMapper().readValue(MaxSpeedCalculator.class.getResource("legal_default_speeds.json"), SpeedLimitsJson.class);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        LegalDefaultSpeeds spLimit = new LegalDefaultSpeeds(data.roadTypesByName, data.speedLimitsByCountryCode);
+
+        List<Map<String, String>> relTags = new ArrayList<>();
+        // You can replace the result of any number of placeholders in a tag filter here (e.g. for name = "urban"),
+        // for example if you have another data source to acquire whether a road is in a built-up area or not.
+        // For those you do not want to replace, simply pass on the result of evaluate as result
+        Function2<String, Function0<Boolean>, Boolean> replacerFunction = (name, eval) -> eval.invoke();
+        AllEdgesIterator iter = graph.getAllEdges();
+        while (iter.next()) {
+            // ISO 3166-1 alpha-2 code optionally concatenated with a ISO 3166-2 code, e.g. "DE", "US" or "BE-VLG"
+            String countryCode = iter.get(countryEnumEncodedValue).getAlpha2();
+            Map<String, String> tags = new HashMap<>();
+            tags.put("rural", iter.get(urbanDensityEnc) == UrbanDensity.RURAL ? "yes" : "no");
```

Comment:
You shouldn't need to do this. This is what the replacerFunction is for:

See https://github.com/westnordost/osm-legal-default-speeds#replacing-placeholders

So, something like...

```
Function2<String, Function0<Boolean>, Boolean> replacerFunction = (name, eval) -> {
    if ("urban".equals(name)) {
        return iter.get(urbanDensityEnc) == UrbanDensity.URBAN;
    }
    if ("rural".equals(name)) {
        return iter.get(urbanDensityEnc) == UrbanDensity.RURAL;
    }
    return eval.invoke();
}
```

(Or, first evaluate the filter normally and if it returns false, then fall back to your data source)

Your solution works too, though. But only because `rural=yes/no` is used in the source data (i.e. in the table in the wiki).